File: Data_Structure/python/ch13_Sort/merge_sort.py
import random
list = []
sorted = []

for i in range(0, 20) :
    list.append(random.randrange(1,100))

for i in range(0, 20) :
    sorted.append(0)

def merge(list, left, mid, right, sorted) :
    i = left
    j = mid + 1
    k = left

    while i <= mid and j <= right :
        if list[i] <= list[j] :
            sorted[k] = list[i]
            k += 1
            i += 1
        else:
            sorted[k] = list[j]
            k += 1
            j += 1

    if i > mid :
        for l in range(j, right +1) :
            sorted[k] = list[l]
            k += 1
    else :
        for l in range(i, mid +1) :
            sorted[k] = list[l]
            k += 1

    list[left:right+1] = sorted[left:right+1]


def merge_sort(list, left, right, sorted) :
    if left < right :
        mid = (left + right) // 2
        merge_sort(list, left, mid, sorted)
        merge_sort(list, mid + 1, right, sorted)
        merge(list, left, mid, right, sorted)

merge_sort(list, 0, 19, sorted)

print(list)

File: Data_Structure/JavaDataStructure/src/chapter14_sorting/MergeSort.java
package chapter14_sorting;

public class MergeSort {

  private static int[] merge(int[] list, int left, int mid, int right) {
    int[] sorted = list.clone();
    int i = left, j = mid + 1, k = left;

    while (i <= mid && j <= right) {
      if (list[i] <= list[j]) sorted[k++] = list[i++];
      else sorted[k++] = list[j++];
    }

    if (i > mid)
      for (int l = j; l <= right; l++)
        sorted[k++] = sorted[l];
    else
      for (int l = i; l <= mid; l++)
        sorted[k++] = list[l];

    return sorted;
  }

  public static int[] sort(int[] list, int left, int right) {
    int[] sorted = list.clone();

    int mid;
    if (left < right) {
      mid = (left + right) / 2;
      sorted = sort(sorted, left, mid);
      sorted = sort(sorted, mid + 1, right);
      sorted = merge(sorted, left, mid, right);
    }

    return sorted;
  }

  public static void main(String[] args) {
    int[] list = {5, 1, 3, 7, 2, 9};

    int[] answer = sort(list, 0, list.length - 1);

    for (int i : answer) {
      System.out.println(i);
    }
  }

}
